x-pack/plugins/lens/public/state_management/custom_middleware.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { delay, finalize, switchMap, tap } from 'rxjs/operators';
+import _, { debounce } from 'lodash';
+import { Dispatch, PayloadAction } from '@reduxjs/toolkit';
+import moment from 'moment';
+import { trackUiEvent } from '../lens_ui_telemetry';
+
+import {
+  waitUntilNextSessionCompletes$,
+  DataPublicPluginStart,
+} from '../../../../../src/plugins/data/public';
+import { setState, LensGetState, LensDispatch } from '.';
+import { LensAppState } from './types';
+import { getResolvedDateRange } from '../lib';
+
+function containsDynamicMath(dateMathString: string) {
+  return dateMathString.includes('now');
+}
+const TIME_LAG_PERCENTAGE_LIMIT = 0.02;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const customMiddleware = (data: DataPublicPluginStart) => (store: any) => {
+  const unsubscribeFromExternalContext = subscribeToExternalContext(
+    data,
+    store.getState,
+    store.dispatch
+  );
+  return (next: Dispatch) => (action: PayloadAction<Partial<LensAppState>>) => {
+    if (action.type === 'app/navigateAway') {
+      unsubscribeFromExternalContext();
+    }
+    // if document was modified or sessionId check if too much time passed to update searchSessionId
+    if (
```

Comment:
time range magic and external subscriptions don't seem to have anything in common - would it make sense to have separate middlewares for them? `external_context_middleware` and `time_range_middleware`